tka/state.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tka
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"golang.org/x/crypto/argon2"
+)
+
+// ErrNoSuchKey is returned if the key referenced by a KeyID does not exist.
+var ErrNoSuchKey = errors.New("key not found")
+
+// State describes Tailnet Key Authority state at an instant in time.
+//
+// State is mutated by applying Authority Update Messages (AUMs), resulting
+// in a new State.
+type State struct {
+	LastAUMHash        []byte   `cbor:"1,keyasint"`
+	DisablementSecrets [][]byte `cbor:"2,keyasint"`
+
+	Keys []Key `cbor:"3,keyasint"`
+}
+
+// GetKey returns the trusted key with the specified KeyID.
+func (s State) GetKey(key KeyID) (Key, error) {
+	for _, k := range s.Keys {
+		if bytes.Equal(k.ID(), key) {
+			return k, nil
+		}
+	}
+
+	return Key{}, ErrNoSuchKey
+}
+
+// Clone makes an independent copy of State.
+//
+// NOTE: There is a difference between a nil slice and an empty
+//       slice for encoding purposes, so an implementation of Clone()
+//       must take care to preserve this.
+func (s State) Clone() State {
```

Comment:
Now that you're in the oss repo you can do this with ./cmd/cloner. Add a line like:

```
//go:generate go run tailscale.com/cmd/cloner -type=State
```

Reply:
I spent a while trying to get this to work, but ultimately I think the assumptions in `cloner` are too different and its not super worthwhile adding in the logic for just this case. Specifically:

1. `cloner` assumes a nil slice and an empty slice are equivalent
2. `cloner` works on pointer receiver (and assumes pointer receiver) methods, whereas we are trying to intentionally pass by value for values which can't be empty